File: packages/mobile/ios/SharedStorage.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(SharedStorage, NSObject)

RCT_EXTERN_METHOD(getString:(NSString *)key
                  appGroup:(NSString *)appGroup
                  resolver:(RCTPromiseResolveBlock)resolver
                  rejecter:(RCTPromiseRejectBlock)rejecter)

RCT_EXTERN_METHOD(setString:(NSString *)key
                  value:(NSString *)value
                  appGroup:(NSString *)appGroup
                  resolver:(RCTPromiseResolveBlock)resolver
                  rejecter:(RCTPromiseRejectBlock)rejecter)

RCT_EXTERN_METHOD(removeKey:(NSString *)key
                  appGroup:(NSString *)appGroup
                  resolver:(RCTPromiseResolveBlock)resolver
                  rejecter:(RCTPromiseRejectBlock)rejecter)

@end

File: packages/mobile/ios/SharedStorage.swift
import Foundation
import React

@objc(SharedStorage)
class SharedStorage: NSObject {

  @objc
  func getString(_ key: String, appGroup: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let userDefaults = UserDefaults(suiteName: appGroup) else {
      rejecter("ERROR", "Failed to access App Group: \(appGroup)", nil)
      return
    }

    let value = userDefaults.string(forKey: key)
    resolver(value)
  }

  @objc
  func setString(_ key: String, value: String, appGroup: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let userDefaults = UserDefaults(suiteName: appGroup) else {
      rejecter("ERROR", "Failed to access App Group: \(appGroup)", nil)
      return
    }

    userDefaults.set(value, forKey: key)
    userDefaults.synchronize()
    resolver(nil)
  }

  @objc
  func removeKey(_ key: String, appGroup: String, resolver: @escaping RCTPromiseResolveBlock, rejecter: @escaping RCTPromiseRejectBlock) {
    guard let userDefaults = UserDefaults(suiteName: appGroup) else {
      rejecter("ERROR", "Failed to access App Group: \(appGroup)", nil)
      return
    }

    userDefaults.removeObject(forKey: key)
    userDefaults.synchronize()
    resolver(nil)
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
